Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,402 @@
+#include "nmos/ocsp_behaviour.h"
+
+#if !defined(_WIN32) || defined(CPPREST_FORCE_HTTP_CLIENT_ASIO)
+#include "pplx/pplx_utils.h" // for pplx::complete_at
+#include "nmos/client_utils.h"
+#include "nmos/model.h"
+#include "nmos/ocsp_settings.h"
+#include "nmos/ocsp_utils.h"
+#include "nmos/random.h"
+#include "nmos/slog.h"
+#include "ssl/ssl_utils.h"
+
+namespace nmos
+{
+    namespace details
+    {
+        struct ocsp_shared_state
+        {
+            load_ca_certificates_handler load_ca_certificates;
+            web::uri base_uri;
+            bool ocsp_service_error;
+            std::vector<uint8_t> ocsp_request;
+
+            // how many seconds before next certificate status request
+            int next_request;
+
+            nmos::details::seed_generator seeder;
+            std::default_random_engine engine;
+            std::unique_ptr<web::http::client::http_client> client;
+
+            explicit ocsp_shared_state(load_ca_certificates_handler load_ca_certificates)
+                : load_ca_certificates(std::move(load_ca_certificates))
+                , ocsp_service_error(false)
+                , next_request(0)
+                , engine(seeder)
+            {}
+        };
+
+        void ocsp_behaviour_thread(nmos::model& model, nmos::experimental::ocsp_settings& ocsp_settings, load_ca_certificates_handler load_ca_certificates, load_server_certificates_handler load_client_certificate, slog::base_gate& gate);
+
+        int half_certificate_expiry_from_now(const std::vector<utility::string_t>& certificate_chains_data, slog::base_gate& gate);
+        std::vector<web::uri> get_ocsp_uris(const std::vector<utility::string_t>& certificates_chain_data, slog::base_gate& gate);
+        std::vector<uint8_t> make_ocsp_request(const std::vector<utility::string_t>& certificate_chains_data, slog::base_gate& gate);
```

Comment:
Why `_data` in the parameter names? How is a certificate distinct from its data?